billing/actions/accounts.py

```diff
@@ -220,3 +228,111 @@ def assign_funds_to_invoice(invoice_id: str) -> bool:
             Charge.objects.create(account_id=account_id, amount=-overpayment, product_code=CREDIT_REMAINING)
 
     return invoice.status == Invoice.PAID
+
+
+def update_accounts_delinquent_status(
+    account_ids: List[int],
+    unpaid_invoices_threshold: Optional[int],
+    days_since_last_unpaid_threshold: Optional[int],
+    currency_amount_threshold_map: Optional[dict],
+) -> Tuple[Dict[int, List[str]], List[int]]:
+    """
+    Mark accounts as delinquent or vice versa when some criteria are accomplished
+    :param account_ids: List of account ids to be evaluated
+    :param unpaid_invoices_threshold: Number of unpaid invoices to consider an user as
+                                      a delinquent
+    :param days_since_last_unpaid_threshold: Days to take into account since the last
+                                             unpaid invoice to consider an user as a
+                                             delinquent
+    :param currency_amount_threshold_map: Balance threshold to consider an user as a
+                                          delinquent.
+                                          Ex: {'CHF': 200, 'EUR': 100, 'NOK': 150}
+    :return: (
+        New delinquent accounts and reason map. Ex: {111: ['Reason why', ...], ...},
+        New complaint accounts ids list
+    )
+    """
+    compliant_accounts = Account.objects.filter(id__in=account_ids, delinquent=False)
+    new_delinquent_accounts_map = {}
+    for account in compliant_accounts:
+        is_delinquent, reason = is_account_delinquent(
+            account,
+            unpaid_invoices_threshold,
+            days_since_last_unpaid_threshold,
+            currency_amount_threshold_map,
+        )
+        if is_delinquent:
+            new_delinquent_accounts_map[account.id] = reason
+
+    new_delinquent_accounts_ids = new_delinquent_accounts_map.keys()
+    account_ids = [
+        account_id for account_id in account_ids
+        if account_id not in new_delinquent_accounts_ids
+    ]
+
+    delinquent_accounts = Account.objects.filter(id__in=account_ids, delinquent=True)
+    complaint_accounts_ids = []
+    for account in delinquent_accounts:
+        is_delinquent, _ = is_account_delinquent(
+            account,
+            unpaid_invoices_threshold,
+            days_since_last_unpaid_threshold,
+            currency_amount_threshold_map,
+        )
+        if not is_delinquent:
+            complaint_accounts_ids.append(account.id)
+
+    Account.objects.filter(
+        id__in=new_delinquent_accounts_map.keys()
+    ).update(delinquent=True)
+    Account.objects.filter(id__in=complaint_accounts_ids).update(delinquent=False)
+
+    return new_delinquent_accounts_map, complaint_accounts_ids
+
+
+def is_account_delinquent(
+    account: Account,
+    unpaid_invoices_threshold: Optional[int],
+    days_since_last_unpaid_threshold: Optional[int],
+    currency_amount_threshold_map: Optional[dict],
+) -> Tuple[bool, List[str]]:
```

Comment:
In my opinion no need for the bool, just the list of strings is enough.
Maybe the function should become compute_account_violations

Reply:
Done